pkg/services/app/add_test.go

```diff
@@ -0,0 +1,209 @@
+package app_test
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/weaveworks/weave-gitops/pkg/flux/fluxfakes"
+	"github.com/weaveworks/weave-gitops/pkg/git/gitfakes"
+	"github.com/weaveworks/weave-gitops/pkg/gitproviders/gitprovidersfakes"
+	"github.com/weaveworks/weave-gitops/pkg/kube"
+	"github.com/weaveworks/weave-gitops/pkg/kube/kubefakes"
+	"github.com/weaveworks/weave-gitops/pkg/services/app"
+)
+
+var (
+	gitClient    *gitfakes.FakeGit
+	fluxClient   *fluxfakes.FakeFlux
+	kubeClient   *kubefakes.FakeKube
+	gitProviders *gitprovidersfakes.FakeGitProviderHandler
+
+	appSrv        app.AppService
+	defaultParams app.AddParams
+)
+
+var _ = BeforeEach(func() {
+	gitClient = &gitfakes.FakeGit{}
+	fluxClient = &fluxfakes.FakeFlux{}
+	kubeClient = &kubefakes.FakeKube{
+		GetClusterNameStub: func() (string, error) {
+			return "test-cluster", nil
+		},
+		GetClusterStatusStub: func() kube.ClusterStatus {
+			return kube.WeGOInstalled
+		},
+	}
+	gitProviders = &gitprovidersfakes.FakeGitProviderHandler{}
+
+	deps := &app.Dependencies{
+		Git:          gitClient,
+		Flux:         fluxClient,
+		Kube:         kubeClient,
+		GitProviders: gitProviders,
+	}
+
+	appSrv = app.New(deps)
+
+	defaultParams = app.AddParams{
+		Url:            "https://github.com/foo/bar",
+		Path:           "./kustomize",
+		Branch:         "main",
+		Dir:            ".",
+		DeploymentType: "kustomize",
+		Namespace:      "wego-system",
+	}
+})
+
+var _ = Describe("Add", func() {
+	It("checks for cluster status", func() {
+		err := appSrv.Add(defaultParams)
+		Expect(err).ShouldNot(HaveOccurred())
+
+		Expect(kubeClient.GetClusterStatusCallCount()).To(Equal(1))
+	})
+
+	It("gets the cluster name", func() {
+		err := appSrv.Add(defaultParams)
+		Expect(err).ShouldNot(HaveOccurred())
+
+		Expect(kubeClient.GetClusterNameCallCount()).To(Equal(1))
+	})
+
+	It("creates and deploys a git secret", func() {
+		fluxClient.CreateSecretGitStub = func(s1, s2, s3 string) ([]byte, error) {
```

Comment:
I'm curious this seems an overwrite. I would suggest  something like `appSrv.SetFluxClient`or `appSrv.GetFluxClient().CreateSecretGitStub` so the intention is more explicit and easier to tell what the options are. Otherwise I need to lookup and the relation between fluxClient and appSrv. I wonder if there is a way to  go back to the previous logic of `CreateSecretGitStub` if needed in the same `It` 🤔 .

Reply:
I think the effect is the same, you need to understand their relationships in either way.

To go back to the previous logic or change the behavior, just set that again, it's just a pointer.